File: content/LLD/Reference-Code/22. Ride Sharing Project - Good Example/ride_matching_service.py
from typing import List
from driver import Driver
from passenger import Passenger
from location import Location
from fare_strategy import FareStrategy
from ride import Ride, RideStatus


class RideMatchingService:
    def __init__(self):
        self.__available_drivers: List[Driver] = []

    def add_driver(self, driver: Driver):
        self.__available_drivers.append(driver)

    def requestRide(
        self, passenger: Passenger, distance: float, strategy: FareStrategy
    ):
        if len(self.__available_drivers) == 0:
            passenger.notify("No Drivers available")
            return

        # Nearest Driver
        nearestDriver: Driver = self.__findNearestDriver(passenger.get_location())
        self.__available_drivers.remove(nearestDriver)

        ride: Ride = Ride(passenger, nearestDriver, distance, strategy)
        ride.calculateFare()

        passenger.notify(f"Ride scheduled with fare Rs.{ride.getRideFare()}")
        nearestDriver.notify(f"You have one new ride for Rs{ride.getRideFare()}")

        ride.updateStatus(RideStatus.ONGOING)

        # After SOmetime
        ride.updateStatus(RideStatus.COMPLETED)
        self.__available_drivers.append(nearestDriver)
        return

    def __findNearestDriver(self, passenger_location: Location) -> Driver:
        assignedDriver = None
        minDistance = float("inf")
        for driver in self.__available_drivers:
            dist = driver.get_location().calcDistance(passenger_location)
            if dist < minDistance:
                minDistance = dist
                assignedDriver = driver
        return assignedDriver

File: content/LLD/Reference-Code/22. Ride Sharing Project - Good Example/RideMatchingService.java
import java.util.ArrayList;
import java.util.List;

public class RideMatchingService {
    private final List<Driver> availableDrivers = new ArrayList<>();

    public void addDriver(Driver driver) {
        this.availableDrivers.add(driver);
    }

    public void requestRide(Passenger passenger, double distance, FareStrategy strategy) {
        if (this.availableDrivers.isEmpty()) {
            passenger.notify("No Drivers available");
            return;
        }

        // Nearest Driver
        Driver nearestDriver = findNearestDriver(passenger.getLocation());
        this.availableDrivers.remove(nearestDriver);

        Ride ride = new Ride(passenger, nearestDriver, distance, strategy);
        ride.calculateFare();

        passenger.notify("Ride scheduled with fare Rs." + ride.getRideFare());
        nearestDriver.notify("You have one new ride for Rs" + ride.getRideFare());

        ride.updateStatus(RideStatus.ONGOING);

        // After some time
        ride.updateStatus(RideStatus.COMPLETED);
        this.availableDrivers.add(nearestDriver);
    }

    private Driver findNearestDriver(Location passengerLocation) {
        Driver assignedDriver = null;
        double minDistance = Double.POSITIVE_INFINITY;
        for (Driver driver : this.availableDrivers) {
            double dist = driver.getLocation().calcDistance(passengerLocation);
            if (dist < minDistance) {
                minDistance = dist;
                assignedDriver = driver;
            }
        }
        return assignedDriver;
    }
}
